### DISCUSSION TITLE
App does not work the same as exe file

### DISCUSSION THREAD

**Discussion by whathellahor** [pyinstaller/pyinstaller#8018] [Help]
Hello everyone, 

I am trying to build an app from a python tkinter program. Basically, this tkinter program will call some shell scripts to perform the program so it references multiple files. I managed to get the whole program to work through the generated unix executable file.

My commands to generate this file would be:

1. I generate the spec file using this:
`pyinstaller path/to/main.py`
2. After the spec file is generated, I added in the relevant datas in the spec file
3. I run pyinstaller against my spec file
`pyinstaller path/to/main.spec`

The above will generate a UNIX executable file, which works. When clicked on it, it will open up a terminal, followed by the tkinter program GUI. I am able to send data, and everything works perfectly. 

However, I want to bundle it into an app, so that the terminal won't show while retaining the functionality like the unix executable file. I did some research and the commands I ran was different. 

1. Generate the spec file using this:
`pyinstaller --onedir --noconsole --clean --windowed -y path/to/main.py`
2. This generates a spec file which I then added the relevant datas
3. Then I run pyinstaller against the spec file again
`pyinstaller path/to/main.spec`

This is where the problem comes. The command will generate a dist folder which contains both UNIX executable file and .app file. Once again, the UNIX executable file works 100% but the .app file does not work at all. 

Does anyone have a fix for this? Thank you in advance!

**Currently using** 
MacOS Ventura 13.5.2
Python 3.11.6
Pyinstaller 6.1.0

**Comment by bwoodsend** [comment]
What if you run your `.app` from the terminal (`./dist/your-app.app/Contents/MacOS/your-app`)? Does that give you an error message you can work with? If it doesn't then you'll need to add some diagnostics handling to your application so that you can see the error message:

```python
import traceback

try:
    your code here
except:
    with open("/Users/home/error.log", "w") as f:
        f.write(traceback.format_exc())
```

Most likely, it'll be due to the reduced environment that desktop applications run in. Anything that runs in a terminal in macOS runs an elaborate sequence of .bashrc and .profile scripts, adding a large number of environment variables which most people are misled into thinking are global environment variables. The big ones are:

* Applications see no custom `PATH` entries. Most importantly, no `/usr/local/bin` or `/opt/homebrew/bin` which means that **your application cannot depend on anything from homebrew** unless you copy it into your application (which is generally a lot harder than it sounds because most homebrew packages will have dependencies).
* `HOME` is not set
* `LC_LANG` is not set so the OS switches from utf8 encodings to ascii.

It's also possible that you're referencing your scripts wrong using relative paths. If you `cd` to a random working directory then try and run your POSIX executable using an absolute path, you should get an error if your path referencing is wrong.
